### PR TITLE
Introduced ImageLoad observer.

### LINKED ISSUE #213: Introduce ImageLoad observer.
We need to fire `ViewDocument#layoutChanged` event every time when `` element has been loaded in the DOM root.

See: https://github.com/ckeditor/ckeditor5-image/issues/142.

### LINKED ISSUE #5151: Introduce ImageLoad observer.
We need to fire `ViewDocument#layoutChanged` event every time when `` element has been loaded in the DOM root.

See: https://github.com/ckeditor/ckeditor5-image/issues/142.

### PR DESCRIPTION
**Pull request by oskarwrobel**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced `ImageLoadObserver`. Closes ckeditor/ckeditor5#5151.

---

### Additional information

*For example – encountered issues, assumptions you had to make, other affected tickets, etc.*

### CODE REVIEW

File: src/image/imageloadobserver.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+	/**
+	 * @inheritDoc
+	 */
+	observe( domRoot, name ) {
+		const viewRoot = this.document.getRoot( name );
+
+		viewRoot.on( 'change:children', ( evt, node ) => {
+			this.view.once( 'render', () => {
+				if ( !node.is( 'element' ) || node.is( 'attributeElement' ) ) {
```
Comment (Reinmar) [inline src/image/imageloadobserver.js]: Extract it out please, cause I can't parse it.

File: src/image/imageloadobserver.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		 */
+		this._observedElements = new Set();
+
+		/**
+		 * DOM emitter used for listening images `load` event.
+		 *
+		 * @private
+		 * @type {module:utils/dom/emittermixin~DomEmitterMixin}
+		 */
+		this._domEmitter = Object.create( DOMEmitterMixin );
```
Comment (Reinmar) [inline src/image/imageloadobserver.js]: `_observer`

Comment (Reinmar) [reply src/image/imageloadobserver.js]: Or `_domObserver`

It doesn't emit anything.

File: src/image/imageloadobserver.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+			} );
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	onDomEvent( domEvent ) {
+		this.document.fire( 'layoutChanged' );
+		this.fire( 'imageLoaded', domEvent );
```
Comment (Reinmar) [inline src/image/imageloadobserver.js]: On `this`? Why not on the document?

**Comment by Reinmar** [comment]
Just a quick review, I haven't checked much, so it may not be all yet.

File: src/image/imageloadobserver.js (hunk 3 above)
Comment (oskarwrobel) [reply src/image/imageloadobserver.js]: https://github.com/ckeditor/ckeditor5-engine/blob/dfa0b39caea05b6df984c6a49949091c8186557c/src/view/observer/domeventobserver.js#L94-L97

Comment (Reinmar) [reply src/image/imageloadobserver.js]: Uh, I forgot that this is confusing.

Comment (Reinmar) [reply src/image/imageloadobserver.js]: https://github.com/ckeditor/ckeditor5-engine/issues/1448 RFC

File: src/image/imageloadobserver.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+ * Observes all new images added to the {@link module:engine/view/document~Document},
+ * fires {@link module:engine/view/document~Document#event:imageLoaded} and
+ * {@link module:engine/view/document~Document#layoutChanged} event every time when the new image
+ * has been loaded.
+ *
+ * **Note:** This event is not fired for images that has been added to the document and rendered as `complete` (already loaded).
+ *
+ * @extends module:engine/view/observer/domeventobserver~DomEventObserver
+ */
+export default class ImageLoadObserver extends DomEventObserver {
```
Comment (pjasiun) [inline src/image/imageloadobserver.js]: Not sure if it should be `DomEventObserver`. You overwrite most of it anyway and the only part you do not overwrite, `fire` method is misleading.

File: src/image/imageloadobserver.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		 */
+		this._observedElements = new Set();
+
+		/**
+		 * DOM emitter used for listening images `load` event.
+		 *
+		 * @private
+		 * @type {module:utils/dom/emittermixin~DomEmitterMixin}
+		 */
+		this._domObserver = Object.create( DOMEmitterMixin );
```
Comment (pjasiun) [inline src/image/imageloadobserver.js]: `Observer` extends `DomEmitterMixin`. Can't you use `this` instead of `_domObserver`?

Comment (oskarwrobel) [reply src/image/imageloadobserver.js]: 🤔I'm sure I checked this and I saw EmitterMixin, not DOMEmitterMixin.

Comment (pjasiun) [reply src/image/imageloadobserver.js]: https://github.com/ckeditor/ckeditor5-engine/blob/dab70e633389b6944b5be38e00dc9511327c5840/src/view/observer/observer.js#L94

Comment (oskarwrobel) [reply src/image/imageloadobserver.js]: `¯\_(ツ)_/¯`
Fixed.

**Pull request merged**